extensions/interactions/GraphInput/GraphInput.js

```diff
@@ -187,362 +191,468 @@ oppia.directive('graphViz', function() {
+          }
+        };
+        $scope.getDirectedEdgeArrowPoints = function(index) {
+          return graphDetailService.getDirectedEdgeArrowPoints(
+            $scope.graph, index);
+        };
+        $scope.getEdgeCentre = function(index) {
+          return graphDetailService.getEdgeCentre($scope.graph, index);
+        };
+      }]
+  };
+});
+
+oppia.factory('graphInputRulesService', [function() {
+  var isIsomophic = function(graph1, graph2) {
+    if (graph1.vertices.length != graph2.vertices.length) {
+      return false;
+    }
+
+    var constructAdjacencyMatrix = function(graph) {
+      var adjMatrix = [];
+      graph.vertices.map(function() {
+        adjMatrix.push([]);
+      });
+      graph.edges.map(function(edge) {
+        var weight = graph.isWeighted ? edge.weight : 1;
+        adjMatrix[edge.src][edge.dst] = weight;
+        if (!graph.isDirected) {
+          adjMatrix[edge.dst][edge.src] = weight;
         }
-        return $scope.graph.vertices[$scope.state.selectedVertex].label;
-      };
-      $scope.selectedEdgeWeight = function(weight) {
-        if ($scope.state.selectedEdge === null) {
-          return '';
+      });
+      return adjMatrix;
+    };
+
+    var adj1 = constructAdjacencyMatrix(graph1);
+    var adj2 = constructAdjacencyMatrix(graph2);
+
+    var nextPermutation = function(permutation) {
+      // Generates (in place) the next lexicographical permuation.
+      // permutation is a permutation of [0...permutation.length]
+
+      // Find the pivot and successor
+      var pivot = null;
+      var successor = null;
+      permutation.reduce(function(previousValue, currentValue, currentIndex) {
+        if (previousValue < currentValue) {
+          pivot = currentIndex - 1;
         }
-        if (angular.isDefined(weight) && angular.isNumber(weight)) {
-          $scope.graph.edges[$scope.state.selectedEdge].weight = weight;
+        if (pivot !== null && currentValue > permutation[pivot]) {
+          successor = currentIndex;
         }
-        return $scope.graph.edges[$scope.state.selectedEdge].weight;
-      };
+        return currentValue;
+      });
+      if (pivot === null) return null;
+      var tmp = permutation[pivot];
+      permutation[pivot] = permutation[successor];
+      permutation[successor] = tmp;
+      permutation = permutation.concat(permutation.splice(pivot + 1).reverse());
+      return permutation;
+    };
 
+    // Check against every permutation of vectices.
+    // The new index of vertex i in graph2 is permutation[i].
+    var numVertices = graph2.vertices.length;
+    var permutation = [];
+    for (var i = 0; i < numVertices; i++) {
+      permutation.push(i);
+    }
+    while (permutation !== null) {
+      var foundIsomorphism = true;
+      if (graph1.isLabeled && graph2.vertices.some(function(vertex, index) {
+        return vertex.label != graph1.vertices[permutation[index]].label;
+      })) {
+        foundIsomorphism = false;
```

Comment:
hmm, then I'd need to repeat the `permutation = nextPermutation(permutation);` line here.